### ISSUE TITLE
:Git commit gets stuck and then errors out

### ISSUE THREAD

**Issue by agorf** [tpope/vim-fugitive#1488]
Hello.

Thanks a lot for vim-fugitive. I'm also a heavy user of 7 other plugins of yours.

I'm using the latest vim-fugitive from master.

```
14:28 in plugged/vim-fugitive on  master took 18s ➜ pwd
/home/agorf/.vim/plugged/vim-fugitive
14:28 in plugged/vim-fugitive on  master ➜ git log -1
commit e06005a742714a4293f1928bca0aca7c9e383383 (HEAD -> master, origin/master, origin/HEAD)
Author: Tim Pope <...>
Date:   Thu Mar 5 08:34:50 2020 -0500

    Fix staging hunk that comes after binary file
    
    Closes https://github.com/tpope/vim-fugitive/issues/1462
```

When I issue `:Git commit` the commit is successful but it seems that Vim is stuck waiting for the Git process to finish(?) When I then go on and hit any key in the output window of `:Git commit` I get the following error:

```
[my-branch-name a49356433c] my commit msg
 1 file changed, 1 insertion(+), 1 deletion(-)
Error detected while processing function fugitive#Resume[3]..<SNR>114_RunWait[11]..<SNR>114_RunSend:
line    5:
E631: ch_sendraw(): write failed
Error detected while processing function fugitive#Resume:
line    3:
E170: Missing :endwhile
```

Not sure if it helps, but before hitting any key, I see the cursor repeatedly switching back and forth between the file buffer and the command output buffer.

Output of `vim --version` in case it helps:

```
Included patches: 1-197, 322, 377-378, 550, 649, 651, 703, 706-707
Extra patches: 8.1.1401, 8.1.1382, 8.1.1368, 8.1.1367, 8.1.1366, 8.1.1365, 8.1.1046, 8.1.0613, 8.1.0547, 8.1.0546, 8.1.0544, 8.1.0540, 8.1.0539, 8.1.0538, 8.1.0506, 8.1.0208, 8.1.0206, 8.1.0205, 8.1.0189, 8.1.0177, 8.1.0067, 8.1.0066
Modified by pkg-vim-maintainers@lists.alioth.debian.org
Compiled by pkg-vim-maintainers@lists.alioth.debian.org
Huge version with GTK3 GUI.  Features included (+) or not (-):
+acl             +file_in_path    +mouse_sgr       +tag_old_static
+arabic          +find_in_path    -mouse_sysmouse  -tag_any_white
+autocmd         +float           +mouse_urxvt     +tcl
+balloon_eval    +folding         +mouse_xterm     +termguicolors
+browse          -footer          +multi_byte      +terminfo
++builtin_terms  +fork()          +multi_lang      +termresponse
+byte_offset     +gettext         -mzscheme        +textobjects
+channel         -hangul_input    +netbeans_intg   +timers
+cindent         +iconv           +num64           +title
+clientserver    +insert_expand   +packages        +toolbar
+clipboard       +job             +path_extra      +user_commands
+cmdline_compl   +jumplist        +perl            +vertsplit
+cmdline_hist    +keymap          +persistent_undo +virtualedit
+cmdline_info    +lambda          +postscript      +visual
+comments        +langmap         +printer         +visualextra
+conceal         +libcall         +profile         +viminfo
+cryptv          +linebreak       -python          +vreplace
+cscope          +lispindent      +python3         +wildignore
+cursorbind      +listcmds        +quickfix        +wildmenu
+cursorshape     +localmap        +reltime         +windows
+dialog_con_gui  +lua             +rightleft       +writebackup
+diff            +menu            +ruby            +X11
+digraphs        +mksession       +scrollbind      -xfontset
+dnd             +modify_fname    +signs           +xim
-ebcdic          +mouse           +smartindent     +xpm
+emacs_tags      +mouseshape      +startuptime     +xsmp_interact
+eval            +mouse_dec       +statusline      +xterm_clipboard
+ex_extra        +mouse_gpm       -sun_workshop    -xterm_save
+extra_search    -mouse_jsbterm   +syntax          
+farsi           +mouse_netterm   +tag_binary
```

And of `apt policy vim-gtk` and `apt policy vim`:

```
vim-gtk:
  Installed: 2:8.1.2269-1
vim:
  Installed: 2:8.0.0197-4+deb9u3
```

I'd be glad to provide more information if necessary.

**Comment by tpope** [comment]
Do commands that don't need to edit a file such as `:Git status` also get stuck? What about *other* commands that edit a file, such as `:Git rebase -i`?

**Comment by tpope** [comment]
Do you have a `post-commit` hook? If so can you try disabling it?

**Comment by agorf** [comment]
> Do you have a `post-commit` hook? If so can you try disabling it?

That was it. I can't believe I checked for any pre-commit hooks and not post-commit.

Thanks, and sorry for wasting your time.

This is probably not the right place for this, but do you accept one-time donations or only sponsorships?

**Comment by tpope** [comment]
If the contents of the `post-commit` hook are something you can share, that might help others experiencing a similar issue. Or perhaps there is a way I can mitigate it.

There's a PayPal link in the sponsor button. You can also sponsor and immediately cancel; that will let you take advantage of GitHub's first year funds matching (i.e., I get double what you put in).

**Comment by agorf** [comment]
Not sure this is a vim-fugitive issue since I've also had issues with this with another tool (fastlane).

Here are the contents:

```bash
#!/bin/sh

git log --format="%H" -n 1 | tr -d '\n' | xclip

exit 0
```

I use it for copying the last commit hash to the clipboard. The only thing that I can see causing the issue is xclip, but I haven't found why.

Wrapping the script with `[[ -t 1 ]]` seems to fix things:

```bash
#!/usr/bin/env bash

if [[ -t 1 ]]; then
  git log --format="%H" -n 1 | tr -d '\n' | xclip
fi

exit 0
```

**Comment by tpope** [comment]
I would expect `[[ -t 1 ]]` to be true in Fugitive, unless you're on Windows (doesn't look like it) or you set the debugging option `g:fugitive_pty = 0`? Did you?

**Comment by agorf** [comment]
Yep, you are right. I'm on Debian GNU/Linux.

It looks like it's false. Just double-checked with:

```bash
if [[ -t 1 ]]; then
  touch ~/itstrue
fi
```
and no file was created after committing from Vim.

Nope, I haven't set `g:fugitive_pty` and `:echo g:fugitive_pty` returns the following in Vim:

```
E121: Undefined variable: g:fugitive_pty
E15: Invalid expression: g:fugitive_pty
```

Not sure that's the right way to inspect the variable though.

**Comment by tpope** [comment]
Oh, right, Vim added PTY support in 8.0.0744, so your version is too old for it.

Poking at `xclip` a bit out of curiosity, stdout is indeed relevant, as `echo test|xclip|cat` indeed hangs. `echo test|xclip -verbose` hangs as well, and gives a clue as to why:

```
Connected to X server.
Using UTF8_STRING.
Waiting for selection requests, Control-C to quit
  Waiting for selection request number 1
```

`xclip` needs to continue running to hold the selection in memory and provide it to applications when paste is invoked. Ordinarily it forks into the background to do this, but for whatever reason, it skips that step when stdout isn't a terminal. My curiosity was exhausted before I was able to figure out why, but not before I found a workaround: `echo test|xclip > /dev/null`.

P.S. Thanks! :wink:

**Comment by agorf** [comment]
You are right. It looks like `-verbose` disabling the forking is intentional: https://stackoverflow.com/questions/19254114/xclip-does-not-terminate-when-tracing-it

Not sure why this happens when stdout is not connected to a terminal though. I tried `> /dev/null` but it still hangs. I tried replacing `xclip` with `xsel` and it looks like it's not experiencing the same issue, so I'm gonna switch to it. The bad thing is if I'm not on a terminal, the commit hash is not correctly copied. But at least, nothing hangs.

Thanks for looking into it and for all of your work with the plugins!
